archinstall/lib/disk/device_model.py

```diff
@@ -0,0 +1,1033 @@
+			mountpoints=mountpoints,
+			btrfs_subvol_infos=btrfs_subvol_infos
+		)
+
+
+@dataclass
+class _DeviceInfo:
+	model: str
+	path: Path
+	type: str
+	total_size: Size
+	free_space_regions: List[DeviceGeometry]
+	sector_size: Size
+	read_only: bool
+	dirty: bool
+
+	def as_json(self) -> Dict[str, Any]:
+		total_free_space = sum([region.get_length(unit=Unit.MiB) for region in self.free_space_regions])
+		return {
+			'Model': self.model,
+			'Path': str(self.path),
+			'Type': self.type,
+			'Size': self.total_size.format_size(Unit.MiB),
+			'Free space': int(total_free_space),
+			'Sector size': self.sector_size.value,
+			'Read only': self.read_only
+		}
+
+	@classmethod
+	def from_disk(cls, disk: Disk) -> _DeviceInfo:
+		device = disk.device
+		device_type = parted.devices[device.type]
+
+		sector_size = Size(device.sectorSize, Unit.B)
+		free_space = [DeviceGeometry(g, sector_size) for g in disk.getFreeSpaceRegions()]
+
+		return _DeviceInfo(
+			model=device.model.strip(),
+			path=Path(device.path),
+			type=device_type,
+			sector_size=sector_size,
+			total_size=Size(int(device.getLength(unit='B')), Unit.B),
+			free_space_regions=free_space,
+			read_only=device.readOnly,
+			dirty=device.dirty
+		)
+
+
+@dataclass
+class SubvolumeModification:
+	name: Path
+	mountpoint: Optional[Path] = None
+	compress: bool = False
+	nodatacow: bool = False
+
+	@classmethod
+	def from_existing_subvol_info(cls, info: _BtrfsSubvolumeInfo) -> SubvolumeModification:
+		return SubvolumeModification(info.name, mountpoint=info.mountpoint)
+
+	@classmethod
+	def parse_args(cls, subvol_args: List[Dict[str, Any]]) -> List[SubvolumeModification]:
+		mods = []
+		for entry in subvol_args:
+			if not entry.get('name', None) or not entry.get('mountpoint', None):
+				log(f'Subvolume arg is missing name: {entry}', level=logging.DEBUG)
+				continue
+
+			mountpoint = Path(entry['mountpoint']) if entry['mountpoint'] else None
+
+			mods.append(
+				SubvolumeModification(
+					entry['name'],
+					mountpoint,
+					entry.get('compress', False),
+					entry.get('nodatacow', False)
+				)
+			)
+
+		return mods
+
+	@property
+	def mount_options(self) -> List[str]:
+		options = []
+		options += ['compress'] if self.compress else []
+		options += ['nodatacow'] if self.nodatacow else []
+		return options
+
+	@property
+	def relative_mountpoint(self) -> Path:
+		"""
+		Will return the relative path based on the anchor
+		e.g. Path('/mnt/test') -> Path('mnt/test')
+		"""
+		if self.mountpoint is not None:
+			return self.mountpoint.relative_to(self.mountpoint.anchor)
+
+		raise ValueError('Mountpoint is not specified')
+
+	def is_root(self, relative_mountpoint: Optional[Path] = None) -> bool:
+		if self.mountpoint:
+			if relative_mountpoint is not None:
+				return self.mountpoint.relative_to(relative_mountpoint) == Path('.')
+			return self.mountpoint == Path('/')
+		return False
+
+	def __dump__(self) -> Dict[str, Any]:
+		return {
+			'name': self.name,
```

Comment:
```suggestion
			'name': str(self.name),
```

Fixes https://github.com/archlinux/archinstall/pull/1604#issuecomment-1510363959